### PR TITLE
#83 - ADS131M04 Driver

### LINKED ISSUE #83: [Driver] - ADS131M04 Driver
### Current Features

We have existing drivers, to reference. This chip communicates over SPI and has an interrupt so we should look at the LTC6820 driver for SPI best practices

### Desired Additional Features

Look at datasheet and implement driver accordingly. https://www.ti.com/lit/ds/symlink/ads131m04-q1.pdf?ts=1708731874014&ref_url=https%253A%252F%252Fwww.ti.com%252Fproduct%252FADS131M04-Q1%252Fpart-details%252FADS131M04QPWRQ1

### Screenshots (as needed)

_No response_

### PR DESCRIPTION
**Pull request by RChandler234**
## Changes

The basic interface for the driver, setting up some basic SPI, and initializing the ADC into continuous-conversion mode using the RESET command. Mainly trying to figure out whether I need to configure anything to get good readings from the ADC and the best way to read out the values generated by the ADC

## Notes
data-ready interrupt signal, DRDY
The end of tPOR is indicated by a transition of DRDY from low to high. The transition of DRDY from low to
high also indicates the SPI interface is ready to accept commands.
Connect the DRDY pin to a digital input on the host to trigger periodic data retrieval in conversion mode.
The conversion status of all channels is available as the DRDY[3:0] bits in the
STATUS register. The STATUS register content is automatically output as the response to the NULL command.
SPI: The device defaults to a 24-bit word size.

 Either strobe the SYNC/RESET pin to re-synchronize conversions and clear the
FIFOs, or quickly read two data packets when data are read for the first time or after a gap in reading data.
This process ensures predictable DRDY pin behavior.

Notes for self from data sheet:
image
image
image
image

Might be important:
image
image
image

## Todo

- [ ] Make sure I'm writing the SPI commands in the right format
- [ ] Make sure everything is configured correctly to get good readings
- [ ] Implement functionality to read out of the ADC

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #83

### CODE REVIEW

File: general/include/ADS131M04.h [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "stm32f4xx_hal.h"
+#include "stm32f4xx_hal_spi.h"
+
+typedef struct
+{
+  SPI_HandleTypeDef *spi;
+  GPIO_TypeDef *gpio;
+  uint8_t cs_pin;
+
+} ads_config;
```
Comment (nwdepatie) [inline general/include/ADS131M04.h:13]: nit: Maybe change to `ads131_t`, might conflict with another similar chip

File: general/include/ADS131M04.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+#define NULL_CMD 0x0
+#define RESET_CMD 0x11
+#define STANDBY_CMD 0x22
+#define WAKEUP_CMD 0x33
+#define LOCK_CMD 0x555
+#define UNLOCK_CMD 0x655
```
Comment (nwdepatie) [inline general/include/ADS131M04.h:20]: Add defines in C file to limit scope

File: general/include/ADS131M04.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+// Method to initialize communication over SPI and
+void ADS131M04_initialize(ltc_config *conf, SPI_HandleTypeDef *hspi, GPIO_TypeDef *hgpio, uint8_t cs_pin);
```
Comment (nwdepatie) [inline general/include/ADS131M04.h:23]: nit: stick with `lower_snake_case`

File: general/src/ADS131M04.c [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+// Function to send command over SPI
+void ADS131M04_send_command(ads_config *config, uint8_t cmd)
+{
+    HAL_GPIO_WritePin(config->gpio, config->cs_pin, GPIO_PIN_RESET);
+    HAL_SPI_Transmit(config->spi, cmd, 1, HAL_MAX_DELAY);
+    HAL_GPIO_WritePin(config->gpio, config->cs_pin, GPIO_PIN_SET);
+}
+
+// Method to abstract writing to a register, will use SPI commands under the hood
```
Comment (nwdepatie) [inline general/src/ADS131M04.c:30]: Nit: I like to use `/* this style */ ` for permanent comments

File: general/include/ADS131M04.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+void ADS131M04_send_command(uint8_t cmd);
+
+// Method to abstract writing to a register, will use SPI commands under the hood
+void ADS131M04_write_reg(uint8_t reg, uint8_t value);
+
+// Method to abstract reading from a register, will use SPI commands under the hood to do
+uint8_t ADS131M04_read_reg(uint8_t reg);
+
+// should read values out of the ADC
+uint8_t ADS131M04_read_ADC();
```
Comment (nwdepatie) [inline general/include/ADS131M04.h:35]: nit: add newline at end of file

File: general/src/ADS131M04.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+// Method to abstract reading from a register, will use SPI commands under the hood to do
+uint8_t ADS131M04_read_reg(uint8_t reg)
+{
+}
+
+// should read values out of the ADC
+uint8_t ADS131M04_read_ADC()
+{
+}
```
Comment (nwdepatie) [inline general/src/ADS131M04.c:43]: nit: add newline at end of file

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Looks solid but you should prob add the actual code that does stuff lol

**Comment by RChandler234** [comment]
image

**Review by nwdepatie — APPROVED** [review]
Made some changes but looks god to me rn, we'll validate it when HV proteus board is assembled

**Pull request merged**